### PR TITLE
(Not for submission) First prototype of BigQuery NUMERIC type support

### PR DESCRIPTION
**Pull request by jskeet**
This involves a new struct, BigQueryNumeric, which is the type
returned from BigQueryRow for NUMERIC vlaues. It has full fidelity,
along with conversions to and from System.Decimal.

Lots more work to do on performance, additional operations (e.g.
division and multiplication), lots more tests.

We may want to accept System.Decimal values for BigQueryParameter
and BigQueryInsertRow, but throw if they lose precision.

### CODE REVIEW

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+            int scale = 9 - exponent;
+            return scale < 0 ? new BigQueryNumeric(rawInteger / s_powersOf10[-scale])
+                : new BigQueryNumeric(rawInteger * s_powersOf10[scale]);
+        }
+
+        // Conversions from CLR types to BigQueryNumeric
+
+        /// <summary>
+        /// Implicit conversion from <see cref="Decimal"/> to <see cref="BigQueryNumeric"/>.
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: "Implicit" => "Explicit"

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        /// </summary>
+        /// <remarks>
+        /// This conversion may silently lose precision. Use <see cref="FromDecimal"/> passing in <see cref="LossOfPrecisionHandling.Throw"/>
+        /// for the second argument to avoid any information loss.
+        /// </remarks>
+        /// <param name="value">The integer value to convert.</param>
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: "integer" => "decimal"

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            if (lossOfPrecisionHandling == LossOfPrecisionHandling.Throw && this != (BigQueryNumeric) dec)
+            {
+                // TODO: Is this the right exception to use?
+                throw new InvalidOperationException("Conversion would lose information");
+            }
+            return dec;
+        }
+
+        /// <summary>
+        /// Implicit conversion from <see cref="BigQueryNumeric"/> to <see cref="Decimal"/>.
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: "Implicit" => "Explicit"

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        /// </summary>
+        /// <remarks>
+        /// This conversion may silently lose precision, but will throw <see cref="OverflowException"/> if the value is out of
+        /// range of <see cref="Decimal"/>. Use <see cref="ToDecimal"/> passing in <see cref="LossOfPrecisionHandling.Throw"/>
+        /// for the second argument to avoid any information loss.
+        /// </remarks>
+        /// <param name="value">The integer value to convert.</param>
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: "integer" => "numeric"?

**Comment by jskeet** [comment]
Thanks for all the comments - will address them over time as this progresses. Loads to do...

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/BigQueryNumericTest.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            BigQueryNumeric numeric = BigQueryNumeric.Parse(text);
+            Assert.Throws<InvalidOperationException>(() => numeric.ToDecimal(LossOfPrecisionHandling.Throw));
+        }
+
+        [Theory]
+        [InlineData("123456789012345678901234567.890123456")]
+        [InlineData("-123456789012345678901234567.890123456")]
+        [InlineData("79228162514264337593543950335.000000001")]
+        [InlineData("5")]
+        [InlineData("-5")]
```
[the remark this answers is not in the document]
Comment (staniscgit) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/BigQueryNumericTest.cs]: Also might want to add numeric max and min values

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+            int exponent = (bits[3] & 0x1f0000) >> 16;
+            if (bits[3] < 0)
+            {
+                rawInteger = -rawInteger;
+            }
+
+            // TODO: All kinds of validation
+
+            int scale = 9 - exponent;
+            return scale < 0 ? new BigQueryNumeric(rawInteger / s_powersOf10[-scale])
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: That's what the TODO is about  :)

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+            {
+                value = default;
+                return "Text representation must be digits, containing an optional decimal point, and an optional leading '-' sign.";
+            }
+            int pointIndex = text.IndexOf('.');
+            if (pointIndex != -1)
+            {
+                int decimalPlaces = text.Length - pointIndex - 1;
+                text = text.Remove(pointIndex, 1);
+                if (decimalPlaces > 9)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: Maybe - elsewhere, .NET *does* generally allow larger values to be parsed. We could add an overload accepting a LossOfPrecisionHandling value though.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Returns a canonical string representation of this value. This always uses "." as a decimal point,
+        /// and only includes as many decimal places as are required to completely represent the value. If
+        /// the value is between -1 and 1 exclusive, a "0" character is included before the decimal point.
+        /// </summary>
+        /// <returns>A canonical string representation of this value.</returns>
+        public override string ToString()
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryNumeric.cs]: *All* of this is written in the quickest way I could get to work. There are string allocations all over the place that we'll want to get rid of. Your suggestion sounds like a good one (or possibly using `Span<T>` in future) but I'd like to nail down the user-visible API surface first.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/BigQueryNumericTest.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/BigQueryNumericTest.cs]: Sure - there are lots of tests to add.

**Comment by jskeet** [comment]
Merging this into a new branch for easier subsequent review.

**Pull request merged**